BUILD.md

```diff
@@ -0,0 +1,359 @@
+```
+Or we can build a custom search path and then use:
+```
+find_file(HAVE_JSONC_H json/json.h PATHS ${OUR_INCLUDE_PATHS})
+```
+
+Note: we may have cases where there is no `.pc` but we have access to a `.cmake` (e.g. AWS SDK, mongodb,cmocka) - these need to be checked / pulled inside the repo while building a prototype.
+
+### Compiler compatability checks
+
+In CMakeLists.txt:
+
+```
+CHECK_INCLUDE_FILE(sys/prctl.h HAVE_PRCTL_H)
+configure_file(cmake/config.in config.h)
+```
+
+In cmake/config.in:
+
+```
+#cmakedefine HAVE_PRCTL_H 1
+```
+
+If we want to check explicitly if something compiles (e.g. the accept4 check, or the 
+`strerror_r` typing issue) then we set the `CMAKE_`... paths and then use:
+```
+check_c_source_compiles(
+    "
+    #include <string.h>
+    int main() { char x = *strerror_r(0, &x, sizeof(x)); return 0; }
+    "
+    STRERROR_R_CHAR_P)
+
+```
+This produces a bool that we can use inside CMake or propagate into the `config.h`.
+
+We can handle the atomic checks with:
+```
+check_c_source_compiles(
+    "
+    int main (int argc, char **argv)
+    {
+      volatile unsigned long ul1 = 1, ul2 = 0, ul3 = 2;
+      __atomic_load_n(&ul1, __ATOMIC_SEQ_CST);
+      __atomic_compare_exchange(&ul1, &ul2, &ul3, 1, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST);
+      __atomic_fetch_add(&ul1, 1, __ATOMIC_SEQ_CST);
+      __atomic_fetch_sub(&ul3, 1, __ATOMIC_SEQ_CST);
+      __atomic_or_fetch(&ul1, ul2, __ATOMIC_SEQ_CST);
+      __atomic_and_fetch(&ul1, ul2, __ATOMIC_SEQ_CST);
+      volatile unsigned long long ull1 = 1, ull2 = 0, ull3 = 2;
+      __atomic_load_n(&ull1, __ATOMIC_SEQ_CST);
+      __atomic_compare_exchange(&ull1, &ull2, &ull3, 1, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST);
+      __atomic_fetch_add(&ull1, 1, __ATOMIC_SEQ_CST);
+      __atomic_fetch_sub(&ull3, 1, __ATOMIC_SEQ_CST);
+      __atomic_or_fetch(&ull1, ull2, __ATOMIC_SEQ_CST);
+      __atomic_and_fetch(&ull1, ull2, __ATOMIC_SEQ_CST);
+      return 0;
+    }
+    "
+    HAVE_C__ATOMIC)
+```
+
+For the specific problem of getting the correct type signature in log.c for the `strerror_r`
+calls we can replicate what we have now, or we can delete this code completely and use a 
+better solution that is documented [here](http://www.club.cc.cmu.edu/~cmccabe/blog_strerror.html).
+To replicate what we have now:
+```
+check_c_source_compiles(
+    "
+    #include <string.h>
+    int main() { char x = *strerror_r(0, &x, sizeof(x)); return 0; }
+    "
+    STRERROR_R_CHAR_P)
+
+check_c_source_compiles(
+    "
+    #include <string.h>
+    int main() { int x = strerror_r(0, &x, sizeof(x)); return 0; }
+    "
+    STRERROR_R_INT)
+
+if("${STRERROR_R_CHAR_P}" OR "${STRERROR_R_INT}")
+    set(HAVE_DECL_STRERROR_R 1)
+endif()
+message(STATUS "Result was ${HAVE_DECL_STRERROR_R}")
+
+```
+
+Note: I did not find an explicit way to select compiler when both `clang` and `gcc` are
+present. We might have an implicit way (like redirecting `cc`) but we should put one in.
+
+
```

Comment:
Why do you have 3 blank lines here?

Reply:
For luck :) I can remove them...

Reply:
Nice, I did not wanna force nobody approve it again! :D